### ISSUE TITLE
Add support for nested arrays/structs in oracles

### ISSUE THREAD

**Issue by LogvinovLeon** [noir-lang/noir#4498]
### Problem

Currently oracles can accept only a list of:
```
pub enum ForeignCallParam {
    Single(Value),
    Array(Vec<Value>),
}
```
which makes it impossible to pass nested arrays or structs. This adds additional burden as arguments need to be serialized before being passed to the oracle.

### Happy Case

Would be nice if oracles would support any type in Noir that is serializable and if Arrays and Structs as well as scalar types would be serialisable 

### Project Impact

Nice-to-have

### Impact Context

I would need to write my own serialisation for nested arrays now because of this

### Workaround

Yes

### Workaround Description

Convert every argument to arrays of fields

### Would you like to submit a PR for this Issue?

None

**Comment by TomAFrench** [comment]
We currently support nested arrays and structs as both arguments and return values in foreign calls, but these must all be flattened at the interface as the ACVM isn't aware of Noir types.

In order to serialise/deserialise these types more easily you can reuse the standard `noirc_abi` crate however. You'll need to write a small ABI definition for each foreign call you're using but that would be simpler than writing your own serialisation logic.

In time we may end up outputting a set of these foreign call ABIs into the build artifact automatically but I don't think that this is going to be a priority for a while tbh.

**Comment by TomAFrench** [comment]
@LogvinovLeon  Just to clarify due to https://github.com/noir-lang/noir/issues/528#issuecomment-1983179769

You can have foreign calls with interfaces like 
```
#[oracle(get_struct)]
unconstrained fn get_struct(size: [[Field; 2]; 2]) -> MyStruct {}
```

today without needing to do any flattening on the Noir side.

**Comment by LogvinovLeon** [comment]
@TomAFrench it seemed to not be working correctly for me.
Here is my Oracle definition:
```
#[oracle(get_proof)]
unconstrained fn get_proof_oracle_1(
    _block_no: Field,
    _address: Address,
    _storage_keys: [Bytes32; 1]
) -> StateAndStorageProof {}

unconstrained fn get_proof_unconstrained_1(
    _block_no: Field,
    _address: Address,
    _storage_keys: [Bytes32; 1]
) -> StateAndStorageProof {
    get_proof_oracle_1(_block_no, _address, _storage_keys)
}
```

`Bytes32` is just `[u8; 32]`

When I run it - here is what the server is getting:
```
[2024-03-07 12:47:44.797 +0100] INFO: Received request
    reqId: "req-f"
    jsonRPCRequest: {
      "method": "get_proof",
      "params": [
        {
          "Single": {
            "inner": "000000000000000000000000000000000000000000000000000000000121eac0"
          }
        },
        {
          "Array": [
            {
              "inner": "00000000000000000000000000000000000000000000000000000000000000a0"
            },
            {
              "inner": "00000000000000000000000000000000000000000000000000000000000000b8"
            },
            {
              "inner": "0000000000000000000000000000000000000000000000000000000000000069"
            },
            {
              "inner": "0000000000000000000000000000000000000000000000000000000000000091"
            },
            {
              "inner": "00000000000000000000000000000000000000000000000000000000000000c6"
            },
            {
              "inner": "0000000000000000000000000000000000000000000000000000000000000021"
            },
            {
              "inner": "000000000000000000000000000000000000000000000000000000000000008b"
            },
            {
              "inner": "0000000000000000000000000000000000000000000000000000000000000036"
            },
            {
              "inner": "00000000000000000000000000000000000000000000000000000000000000c1"
            },
            {
              "inner": "00000000000000000000000000000000000000000000000000000000000000d1"
            },
            {
              "inner": "000000000000000000000000000000000000000000000000000000000000009d"
            },
            {
              "inner": "000000000000000000000000000000000000000000000000000000000000004a"
            },
            {
              "inner": "000000000000000000000000000000000000000000000000000000000000002e"
            },
            {
              "inner": "000000000000000000000000000000000000000000000000000000000000009e"
            },
            {
              "inner": "00000000000000000000000000000000000000000000000000000000000000b0"
            },
            {
              "inner": "00000000000000000000000000000000000000000000000000000000000000ce"
            },
            {
              "inner": "0000000000000000000000000000000000000000000000000000000000000036"
            },
            {
              "inner": "0000000000000000000000000000000000000000000000000000000000000006"
            },
            {
              "inner": "00000000000000000000000000000000000000000000000000000000000000eb"
            },
            {
              "inner": "0000000000000000000000000000000000000000000000000000000000000048"
            }
          ]
        },
        {
          "Array": [
            {
              "inner": "0000000000000000000000000000000000000000000000000000000000000055"
            }
          ]
        }
      ],
      "id": 1,
      "jsonrpc": "2.0"
    }
```

So as you see - first and second args are ok, but the third array gets serialized as single value `0000000000000000000000000000000000000000000000000000000000000055`. I don't know where does it come from as I don't pass anything that looks like it. Maybe it's some sort of length?
Here is my `Prover.toml`
```
block_no = "19000000"
address = [
    160,
    184,
    105,
    145,
    198,
    33,
    139,
    54,
    193,
    209,
    157,
    74,
    46,
    158,
    176,
    206,
    54,
    6,
    235,
    72,
]
keys = [
    [
        87,
        209,
        138,
        247,
        147,
        215,
        48,
        12,
        75,
        164,
        109,
        25,
        46,
        199,
        170,
        9,
        80,
        112,
        221,
        230,
        197,
        44,
        104,
        124,
        109,
        13,
        146,
        251,
        133,
        50,
        179,
        5,
    ],
]

```

**Comment by TomAFrench** [comment]
Ah, that's a problem. Looks like we're not reading nested arrays from brillig memory properly, I think the value you're getting out is the pointer to the inner array whereas we should recursively follow these.

**Comment by LogvinovLeon** [comment]
I'm not trying to rush you, but could you please let me know - if it's reasonable to expect it fixed in nightly within a week? Otherwise - we would change our interfaces for now

**Comment by LogvinovLeon** [comment]
I've ran your example and I'm getting:
```
{
      "method": "get_proof",
      "params": [
        {
          "Single": {
            "inner": "0000000000000000000000000000000000000000000000000000000000000000"
          }
        },
        {
          "Single": {
            "inner": "0000000000000000000000000000000000000000000000000000000000000042"
          }
        },
        {
          "Array": [
            {
              "inner": "000000000000000000000000000000000000000000000000000000000000000b"
            },
            {
              "inner": "0000000000000000000000000000000000000000000000000000000000000010"
            }
          ]
        }
      ],
      "id": 1,
      "jsonrpc": "2.0"
    }
```

Which looks like two Brillig array pointers.
Do you run an HTTP Server or do you print from within noir code somewhere? Maybe there is some code after you print?

**Comment by LogvinovLeon** [comment]
I'll try to give you some easy setup with logging HTTP server, but we already have a repro on Noir side

**Comment by TomAFrench** [comment]
I'm getting the results

```
Server listening at http://localhost:5555
Received POST request on /
Payload: {
  "method": "get_proof",
  "params": [
    {
      "Single": {
        "inner": "0000000000000000000000000000000000000000000000000000000000000000"
      }
    },
    {
      "Single": {
        "inner": "0000000000000000000000000000000000000000000000000000000000000042"
      }
    },
    {
      "Array": [
        {
          "inner": "0000000000000000000000000000000000000000000000000000000000000001"
        },
        {
          "inner": "0000000000000000000000000000000000000000000000000000000000000002"
        },
        {
          "inner": "0000000000000000000000000000000000000000000000000000000000000003"
        },
        {
          "inner": "0000000000000000000000000000000000000000000000000000000000000004"
        },
        {
          "inner": "0000000000000000000000000000000000000000000000000000000000000005"
        },
        {
          "inner": "0000000000000000000000000000000000000000000000000000000000000006"
        }
      ]
    }
  ],
  "id": 1,
  "jsonrpc": "2.0"
}

```

**Comment by LogvinovLeon** [comment]
```
nargo version = 0.23.0
noirc version = 0.23.0+5be9f9d7e2f39ca228df10e5a530474af0331704
(git version hash: 5be9f9d7e2f39ca228df10e5a530474af0331704, is dirty: false)
```

What's your nargo version?

**Comment by LogvinovLeon** [comment]
I'll try with nightly now

**Comment by LogvinovLeon** [comment]
Nightly works. Ok, thanks and sorry for wasting your time. I should have checked with nightly first. This is awesome!

**Comment by TomAFrench** [comment]
I'm building off of master

**Comment by TomAFrench** [comment]
No worries, glad that we've got it working for you :)

**Comment by LogvinovLeon** [comment]
FYI: 0.24.0 works, so no need to use nightly here
Do I understand correctly that because of resources limitations - noir doesn't do backports of bug fixes for now as we shouldn't expect it to be fixed in 0.23.0?

**Comment by TomAFrench** [comment]
Yep, we're not doing backports of fixes currently as things are moving too quickly to really provide a ton of guarantees.

**Comment by LogvinovLeon** [comment]
Unfortunately for now we can't use nightly nargo because of this issue: https://github.com/noir-lang/noir/issues/4443
TL;DR Latest nargo can't verify it's own proofs
Also - nightly nargo versions introduced a couple of breaking changes that broke both of our dependencies and the fixes are non-trivial. So for now we'll stay on nargo 0.23.0 and just don't use nested arrays in oracles.
